### PR TITLE
Feature/run programs

### PR DESCRIPTION
**Pull request by belfortep**
## Summary
This PR add new ways to run different programs and an interactive console, also disables the input buffering

## Changes
- Added a Makefile to easier run the virtual machine
- Added the possibility to debug programs and an interactive console, both can be run with `make debug` or `make interactive`
- Disabled the input buffering to omit pressing enter when using the example files

### Minor changes
- updated the README with the new executions methods and instructions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,12 @@
...
 Implementation of lc3-vm in Rust
+
+## Execution
+- `make run FILE=file_name` to run an object file, `file_name` contains the program to execute, else use `make run` to use the example `2048.obj` file.
+- `make interactive` to open an interactive console of the Virtual Machine, where all the instructions you type in binary form will be executed at the moment, also if you type "r", you can see the state of the registers.
+- `make debug FILE=file_name` to run an object file in debug mode, `file_name` contains the program to debug. If you just use `make debug`, it will use the `2048.obj` by default. In another console you have to type `make debugger` to use the debugger there
+
+### Debugger Controls
+- "n" to pass one instruction and see the instruction executed on the debugger
+- "Any u16 number" to pass that ammount of instructions (for example, just type 10000 to pass 10000 instructions forward, recommended for the `2048.obj`)
```
Comment (juanimedone) [inline README.md]:
```suggestion
- "Any u16 number" to execute that amount of instructions (for example, just type 10000 to pass 10000 instructions forward, recommended for the `2048.obj`)
```

File: src/main.rs
Comment (juanimedone) [inline src/main.rs:1]: I'm not sure if it's 100% necessary, but you should restore input buffering when the program finishes.

File: src/command_line_parser/parser.rs
Comment (juanimedone) [inline src/command_line_parser/parser.rs:1]: This is probably not the right name for this file, as it contains functions that do things that are not related to parsing.

File: src/virtual_machine/lc3_virtual_machine.rs [hunk 2]
```diff
@@ -188,6 +188,16 @@ impl LC3VirtualMachine {
         self.decode_instruction(instruction);
     }
 
+    pub fn next_instructions(&mut self, ammount_of_instructions: u16) {
+        for _ in 0..ammount_of_instructions {
```
Comment (juanimedone) [inline src/virtual_machine/lc3_virtual_machine.rs]:
```suggestion
    pub fn next_instructions(&mut self, amount_of_instructions: u16) {
        for _ in 0..amount_of_instructions {
```

File: src/virtual_machine/lc3_virtual_machine.rs [hunk 3]
```diff
@@ -188,6 +188,16 @@ impl LC3VirtualMachine {
...
+        }
+    }
+
+    pub fn state_of_registers(&mut self) -> String {
+        self.registers.state_of_registers()
+    }
+
     fn sign_extend(mut value_to_extend: u16, ammount_of_bits: u16) -> u16 {
         if (value_to_extend >> (ammount_of_bits - 1) & 0b1) == 1 {
             value_to_extend |= 0xFFFF << ammount_of_bits;
```
Comment (juanimedone) [inline src/virtual_machine/lc3_virtual_machine.rs]:
```suggestion
    fn sign_extend(mut value_to_extend: u16, amount_of_bits: u16) -> u16 {
        if (value_to_extend >> (amount_of_bits - 1) & 0b1) == 1 {
            value_to_extend |= 0xFFFF << amount_of_bits;
```

File: src/command_line_parser/parser.rs [hunk 4]
```diff
@@ -23,6 +36,87 @@ pub fn execute_program_from_file(file: &str) -> Result<(), String> {
...
+                writeln!(&stream, "{}", response).map_err(|error| error.to_string())?;
+            }
+            "r" => {
+                let response = virtual_machine.state_of_registers();
+                writeln!(&stream, "{}", response).map_err(|error| error.to_string())?;
+            }
+            _ => match command.parse::<u16>() {
+                Ok(ammount_of_instructions) => {
+                    virtual_machine.next_instructions(ammount_of_instructions);
+                    let response = format!("executed {} instructions", ammount_of_instructions);
```
Comment (juanimedone) [inline src/command_line_parser/parser.rs]:
```suggestion
                Ok(amount_of_instructions) => {
                    virtual_machine.next_instructions(amount_of_instructions);
                    let response = format!("executed {} instructions", amount_of_instructions);
```

**Review by juanimedone — COMMENTED** [review]
Great work on adding the interactive and debug modes—they look fantastic! I've left a few comments (most of them are just of typos), the rest looks good to me. Remember to address the remaining Clippy warning.

File: src/main.rs
Comment (belfortep) [reply src/main.rs:1]: Just tested it with the macOS and vscode terminal and it returns to normal, I think rust just do this when termios struct goes out of scope

File: src/command_line_parser/parser.rs
Comment (belfortep) [reply src/command_line_parser/parser.rs:1]: Now that you mention it, that’s right, it was just a parser but now it evolved to something bigger, maybe having another file for the not parse related functions

File: src/bin/debugger.rs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    println!("Instructions: ");
+    println!("<n> to execute one instruction");
+    println!("<a number> to execute that number of instructions (recommended)");
+    println!("<r> to print the state of the registers of the program");
+    println!("remember to use the program you are debugging if it needs user input");
+}
+
+fn main() -> Result<(), String> {
+    let connection_address = format!("{}:{}", LOCAL_HOST, PORT);
+    let stream = TcpStream::connect(connection_address).map_err(|error| error.to_string())?;
```
Comment (Oppen) [inline src/bin/debugger.rs]: Given you're always connecting only to localhost, doesn't it make more sense to use an IPC socket instead?

Comment (Oppen) [reply src/bin/debugger.rs]: Also, maybe a datagram oriented protocol makes more sense here.

File: src/main.rs
Comment (lazcanoluca) [reply src/main.rs:1]: One of the reviewers gave me the following comment with respect to this:

```
// Having `setup_terminal` and `restore_terminal` means that you are responsible for calling them.
// If you write a `TermiosHandler` newtype struct (that has the `Termios` object) you can move
// `setup_terminal` into `fn new() -> Self` and add an `impl Drop for TermiosHandler` where you can
// put `restore_terminal` so that it gets called automatically when the instance goes out of scope.
```

I think it's very good, as you're not forced to call the restoration of the settings of the terminal yourself, you can just rely on Rust calling `drop` with the cleanup function when the handler goes out of scope. Hope it helps both of you @juanimedone @belfortep .

File: src/virtual_machine/lc3_virtual_machine.rs [hunk 6]
```diff
@@ -188,9 +188,19 @@ impl LC3VirtualMachine {
         self.decode_instruction(instruction);
     }
 
-    fn sign_extend(mut value_to_extend: u16, ammount_of_bits: u16) -> u16 {
-        if (value_to_extend >> (ammount_of_bits - 1) & 0b1) == 1 {
-            value_to_extend |= 0xFFFF << ammount_of_bits;
+    pub fn next_instructions(&mut self, amount_of_instructions: u16) {
```
Comment (Oppen) [inline src/virtual_machine/lc3_virtual_machine.rs]: Maybe steps is a brief but clear name:
```suggestion
    pub fn next_instructions(&mut self, steps: u16) {
```

File: src/virtual_machine_start.rs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                    println!("instruction proccess: {}", instruction);
+                }
+                Err(_) => println!("Wrong instruction format"),
+            },
+        }
+    }
+    Ok(())
+}
+
+fn receive_file(arg: String) -> Result<BufReader<File>, String> {
```
Comment (Oppen) [inline src/virtual_machine_start.rs]: If you change this to:
```suggestion
fn receive_file(arg: impl AsRef<Path>) -> Result<BufReader<File>, String> {
```
You can avoid all the `.to_string` calls you add on the call site. Both `String`, `&str` and `Path` would just work.

File: src/main.rs
Comment (belfortep) [reply src/main.rs:1]: Oh, didn't know about the `Drop` trait, going to take a look at that to handle the restoration of the terminal

File: src/virtual_machine_start.rs (hunk 7 above)
Comment (belfortep) [reply src/virtual_machine_start.rs]: Just searched about the `AsRef` trait, didn't know about it and seems really interesting for generics

File: src/command_line_parser/parser.rs
Comment (juanimedone) [reply src/command_line_parser/parser.rs:1]: Yes, that's a good idea.

File: src/main.rs [hunk 8]
```diff
@@ -1,13 +1,47 @@
...
+}
+
+impl TermiosWrapper {
+    pub fn new(mut termios: Termios) -> Result<Self, String> {
+        termios.c_lflag &= !ICANON & !ECHO;
+
+        tcsetattr(STDIN, TCSANOW, &termios).map_err(|error| error.to_string())?;
+        Ok(Self { termios })
+    }
+}
```
Comment (juanimedone) [inline src/main.rs]: You have to 'save' the original `termios` unmodified in order to restore it in `drop`.
```suggestion
    pub fn new(termios: Termios) -> Result<Self, String> {
        let mut new_termios = termios;
        new_termios.c_lflag &= !ICANON & !ECHO;

        tcsetattr(STDIN, TCSANOW, &new_termios).map_err(|error| error.to_string())?;
        Ok(Self { termios })
    }
}
```

File: src/virtual_machine_start.rs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+fn load_reader_file_to_vm_memory(mut reader: BufReader<File>) -> Result<LC3VirtualMachine, String> {
+    let program_counter_start = reader
+        .read_u16::<BigEndian>()
+        .map_err(|error| error.to_string())?;
+
+    let mut virtual_machine = LC3VirtualMachine::new(program_counter_start);
+    let mut memory_address = program_counter_start;
+    while let Ok(instruction) = reader.read_u16::<BigEndian>() {
+        virtual_machine.memory_write(memory_address, instruction);
+        memory_address += 1;
```
Comment (juanimedone) [inline src/virtual_machine_start.rs]:
```suggestion
        memory_address = memory_address.wrapping_add(1);
```
To prevent `u16` overflow.

File: src/bin/debugger.rs (hunk 5 above)
Comment (belfortep) [reply src/bin/debugger.rs]: Never tried using the IPC unix sockets before and really makes sense to use them in this case for what I read, but why the datagrams ones? It was a nice practice as it was the first time I used `UDP` sockets, but the codes feels way harder to understand (probably the handling of that type of connections that I made wasn't very optimal)

File: src/virtual_machine_start.rs (hunk 9 above)
Comment (belfortep) [reply src/virtual_machine_start.rs]: Oops, missed that one

File: src/main.rs (hunk 8 above)
Comment (belfortep) [reply src/main.rs]: Oh really? Didn't know about that one, thanks 🤝

Comment (belfortep) [reply src/main.rs]: Just tried it and if It's a new termios is created, the code doesn't really works (I tried a lot of things and seems that the moment the struct its created, its also is dropped)

Comment (juanimedone) [reply src/main.rs]: Is it fixed now with [this commit](https://github.com/belfortep/lc3-vm/pull/2/commits/9b5a9c380d0dd3f500108d9e6e6e993132d6af57)?

File: README.md [hunk 10]
```diff
@@ -10,3 +10,6 @@ Implementation of lc3-vm in Rust
 - "n" to pass one instruction and see the instruction executed on the debugger
 - "Any u16 number" to execute that amount of instructions (for example, just type 10000 to pass 10000 instructions forward, recommended for the `2048.obj`)
 - "r" to see the state of the registers
+
+### Recommendation
+If you try the debugger, it will generate a .socket file in your /tmp/ folder. You will need to remove them if you use the debugger more than once. Try using `ls /tmp/*socket` and after that execute the `rm /tmp/*socket` command
```
Comment (Oppen) [inline README.md]: Maybe the debugger/application can remove it on startup.

Comment (belfortep) [reply README.md]: Oh, you are right, just added that

**Review by juanimedone — APPROVED** [review]

File: src/virtual_machine/register.rs [hunk 11]
```diff
@@ -48,6 +48,14 @@ impl Registers {
...
         }
     }
+
+    pub fn state_of_registers(&mut self) -> String {
+        self.registers
+            .iter()
+            .fold(String::new(), |accumulator, register| {
+                accumulator + &format!("{register:#018b}::")
+            })
+    }
```
Comment (lazcanoluca) [inline src/virtual_machine/register.rs]: If I understand correctly, you are using `state_of_registers` as a way to serialize/format the registers, to be displayed in the debugger, right?

I think it would be much more idiomatic to `impl Display for Register`. That way, you get the `.to_string()` method for sending to the debugger.

**Review by Oppen — APPROVED** [review]

File: src/bin/debugger.rs (hunk 5 above)
Comment (Oppen) [reply src/bin/debugger.rs]: I think more in how the message is meant to be interpreted, the debugger is using discrete commands rather than a stream of data from my POV, so you avoid having to keep track of separators and what not. Each message is always a complete command and nothing more, and each message has an associated size.
But you're right about the code seeming harder, it may make sense to revert to stream.

File: src/virtual_machine/register.rs (hunk 11 above)
Comment (belfortep) [reply src/virtual_machine/register.rs]: That makes sense, it's easier to understand that way

File: src/virtual_machine/register.rs [hunk 12]
```diff
@@ -50,6 +50,16 @@ impl Registers {
...
+impl ToString for Registers {
+    fn to_string(&self) -> String {
+        self.registers
+            .iter()
+            .fold(String::new(), |accumulator, register| {
+                accumulator + &format!("{register:#018b}::")
+            })
+    }
+}
+
```
Comment (lazcanoluca) [inline src/virtual_machine/register.rs]: The Rust documentation discourages the use of `ToString`, encouraging the use of `Display` instead, through which you get `.to_string()` for free.

https://doc.rust-lang.org/std/string/trait.ToString.html

> This trait is automatically implemented for any type which implements the [Display](https://doc.rust-lang.org/std/fmt/trait.Display.html) trait. As such, ToString shouldn’t be implemented directly: [Display](https://doc.rust-lang.org/std/fmt/trait.Display.html) should be implemented instead, and you get the ToString implementation for free.

**Review by lazcanoluca — APPROVED** [review]
Left a nitpick, but approving. Lgtm 🚀

File: src/bin/debugger.rs (hunk 5 above)
Comment (belfortep) [reply src/bin/debugger.rs]: I will have that in mind, usually I just go with `TCP` for convenience whenever there are sockets, but knowing the size to avoid separators with `UDP` is interesting

File: src/virtual_machine/register.rs (hunk 12 above)
Comment (belfortep) [reply src/virtual_machine/register.rs]: Oh, just tried to implement `Display` and read that the function was `fmt` and didn't use, now it makes more sense
